### DISCUSSION TITLE
Some questions about using @overload decorator correctly

### DISCUSSION THREAD

**Discussion by szw0407** [microsoft/pylance-release#5168] [Enhancement]
Below is an example code:
```python
from typing import overload

class A:
    data: list[int]  # A
    def __init__(self, *args: int) -> None:
        self.data = list(args)

class B:
    @overload
    def __init__(self, *args: int) -> None: ...
    @overload
    def __init__(self, *args: A) -> None: ...

    def __init__(self, *args: int | A) -> None:
        self.data = []
        if all(isinstance(arg, A) for arg in args):
            for arg in args:
                self.data.extend(arg.data) # typeerror here, saying arg is Unknown type, but arg has been checked as A in the if statement, thus arg.data should be list[int] according to the line marked A
        elif all(isinstance(arg, int) for arg in args):
            self.data.extend(args)  # typeerror here, saying args is Unknown type, but it has been checked as list[int] in the if statement
        else:
            raise TypeError
```

May I give another example code:

```python
from typing import overload, Literal
    
class Obj:
    data: list[int]
    @overload
    def __init__(self, *args: int, all_int: Literal[True]) -> None: ...
    @overload
    def __init__(self, *args: str, all_str: Literal[True]) -> None: ...

    def __init__(self, *args:int | str, **kwargs: bool) -> None:
        if kwargs.keys() == {'all_int'}:
            self.data = list(args) # TypeError: list[int | str] is incompatible with list[int]
        elif kwargs.keys() == {'all_str'}:
            self.data = list(map(int, args))
        else:
            raise TypeError("Invalid arguments")
```

I have declared the Type of args when `all_int` is in kwargs.keys() the args must be all integers, but when `all_str` appears, it will be all strings. I can use `all(isinstance(args, int))` to check the types but I guess that was not necessary if Pylance could read the overload function declearations. Moreover, even if I did so, `all` function is not supported yet.

It might be asked why I am using `kwargs` instead of writing all the names of variables in the function body. Well, I think using `kwargs` makes my code more readable and clear.

Perhaps I misunderstood what overload could do. And I found #4526 . It had been marked as completed, and I am not sure whether some exceptions still exist. Thanks for your help.

**Comment by szw0407** [comment]
The previous version of this question is fixed when I switch to the pre-release version of Pylance, but perhaps there are still some issues.

The new example may be due to usage of the `all` function.

**Reply by erictraut** [reply]
The `all` function is not a supported type guard. For a complete list of supported type guard patterns, please refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards). For complex type guards like `all`, you would need to write your own "user-defined type guard" using the `TypeGuard` capability introduced in [PEP 647](https://peps.python.org/pep-0647/).

**Reply by erictraut** [reply]
Also, it's important to understand that `@overload` signatures do not affect the type checking of the implementation of the overload. The `@overload` signatures are used only when evaluating calls of the function, not the implementation of the function. For more information about overloads, refer to [this section of PEP 484](https://peps.python.org/pep-0484/#function-method-overloading).
